mapviz_plugins/src/object_plugin.cpp

```diff
@@ -0,0 +1,467 @@
+    data.transformed = false;
+
+    swri_transform_util::Transform transform;
+    if (GetTransform(data.source_frame, data.stamp, transform))
+    {
+      data.transformed = true;
+    }
+
+    data.polygon.reserve(obj.polygon.size()+1);
+    for (auto& point: obj.polygon)
+    {
+      tf::Vector3 pt(point.x, point.y, point.z);
+      pt = data.local_transform*pt;
+      tf::Vector3 transformed = transform *pt;
+      data.polygon.push_back({pt, transformed});
+    }
+    if (data.polygon.size())
+    {
+      data.polygon.push_back(data.polygon.front());
+    }
+    objects_.push_back(data);
+  }
+
+  void ObjectPlugin::PrintError(const std::string& message)
+  {
+    PrintErrorHelper(ui_.status, message);
+  }
+
+  void ObjectPlugin::PrintInfo(const std::string& message)
+  {
+    PrintInfoHelper(ui_.status, message);
+  }
+
+  void ObjectPlugin::PrintWarning(const std::string& message)
+  {
+    PrintWarningHelper(ui_.status, message);
+  }
+
+  QWidget* ObjectPlugin::GetConfigWidget(QWidget* parent)
+  {
+    config_widget_->setParent(parent);
+
+    return config_widget_;
+  }
+
+  bool ObjectPlugin::Initialize(QGLWidget* canvas)
+  {
+    canvas_ = canvas;
+    SetColor(ui_.color->color());
+
+    return true;
+  }
+
+  void ObjectPlugin::Draw(double x, double y, double scale)
+  {
+    auto objIter = objects_.begin();
+    while (objIter != objects_.end())
+    {
+      const ObjectData& obj = *objIter;
+
+      if (!obj.transformed) {
+        objIter++;
+        continue;
+      }
+
+      glColor4f(color_.redF(), color_.greenF(), color_.blueF(), 1.0);
+
+      //glColor4f(marker.color.r, marker.color.g, marker.color.b, marker.color.a);
+
+      glLineWidth(3.0);//marker.scale_x);
+      glBegin(GL_LINE_STRIP);
+
+      for (const auto &point : obj.polygon) {
+        //glColor4f(point.color.r, point.color.g, point.color.b, point.color.a);
+
+        glVertex2d(
+          point.transformed_point.getX(),
+          point.transformed_point.getY());
+      }
+
+      glEnd();
+
+      objIter++;
+      PrintInfo("OK");
+    }
+  }
+
+  void ObjectPlugin::Paint(QPainter* painter, double x, double y, double scale)
+  {
+    if (!ui_.show_ids->isChecked())
+    {
+      return;
+    }
+
+    // Most of the marker drawing is done using OpenGL commands, but text labels
+    // are rendered using a QPainter.  This is intended primarily as an example
+    // of how the QPainter works.
+    ros::Time now = ros::Time::now();
+
+    // We don't want the text to be rotated or scaled, but we do want it to be
+    // translated appropriately.  So, we save off the current world transform
+    // and reset it; when we actually draw the text, we'll manually translate
+    // it to the right place.
+    QTransform tf = painter->worldTransform();
+    QFont font("Helvetica", 10);
+    painter->setFont(font);
+    painter->save();
+    painter->resetTransform();
+
+    for (auto objIter = objects_.begin(); objIter != objects_.end(); ++objIter)
```

Comment:
Seems like the loops here and in Transform below could be for-each loops as well.